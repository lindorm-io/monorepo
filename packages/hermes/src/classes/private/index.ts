export * from "./HermesMetadata";
export * from "./HermesRegistry";
export * from "./HermesScanner";
