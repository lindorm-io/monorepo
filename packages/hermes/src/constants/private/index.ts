export * from "./checksum-store";
export * from "./domain";
export * from "./event-store";
export * from "./saga-store";
export * from "./view-store";
