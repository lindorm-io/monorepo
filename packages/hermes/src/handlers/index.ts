export * from "./HermesAggregateCommandHandler";
export * from "./HermesAggregateErrorHandler";
export * from "./HermesAggregateEventHandler";
export * from "./HermesSagaErrorHandler";
export * from "./HermesSagaEventHandler";
export * from "./HermesSagaIdHandler";
export * from "./HermesSagaTimeoutHandler";
export * from "./HermesViewErrorHandler";
export * from "./HermesViewEventHandler";
export * from "./HermesViewIdHandler";
export * from "./HermesViewQueryHandler";
