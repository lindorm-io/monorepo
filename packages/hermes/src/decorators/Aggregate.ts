import { Constructor } from "@lindorm/types";
import { AggregateDecoratorOptions } from "../types";
import { extractNameData, globalHermesMetadata } from "../utils/private";

export function Aggregate(options: AggregateDecoratorOptions = {}): ClassDecorator {
  return function (target) {
    const { name } = extractNameData(target.name);
    globalHermesMetadata.addAggregate({
      encryption: options.encryption || false,
      name: options.name || name,
      namespace: options.namespace || null,
      target: target as unknown as Constructor,
    });
  };
}
