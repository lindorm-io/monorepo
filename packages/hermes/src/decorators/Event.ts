import { Constructor } from "@lindorm/types";
import { EventDecoratorOptions } from "../types";
import { extractNameData, globalHermesMetadata } from "../utils/private";

export function Event(options?: EventDecoratorOptions): ClassDecorator {
  return function (target) {
    const { name, version } = extractNameData(target.name);
    globalHermesMetadata.addEvent({
      name: options?.name || name,
      target: target as unknown as Constructor,
      version: options?.version || version,
    });
  };
}
