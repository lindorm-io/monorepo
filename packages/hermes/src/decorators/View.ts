/* eslint-disable @typescript-eslint/no-unsafe-function-type */

import { isArray } from "@lindorm/is";
import { Constructor } from "@lindorm/types";
import { ViewDecoratorOptions, ViewStoreSource } from "../types";
import { extractNameData, globalHermesMetadata } from "../utils/private";

export function View(
  AggregateClass: Constructor | Array<Constructor>,
  source: ViewStoreSource,
  options: ViewDecoratorOptions = {},
): ClassDecorator {
  return function (target) {
    const { name } = extractNameData(target.name);
    globalHermesMetadata.addView({
      aggregates: isArray(AggregateClass) ? AggregateClass : [AggregateClass],
      name: options.name || name,
      namespace: options.namespace || null,
      source,
      target: target as unknown as Constructor,
    });
  };
}
