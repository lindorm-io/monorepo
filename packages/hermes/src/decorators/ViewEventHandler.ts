import { Constructor, Dict } from "@lindorm/types";
import { ViewEventHandlerDecoratorOptions, ViewEventHandlerDescriptor } from "../types";
import { globalHermesMetadata } from "../utils/private";

export function ViewEventHandler<C extends Constructor, S extends Dict>(
  EventClass: C,
  options: ViewEventHandlerDecoratorOptions = {},
): ViewEventHandlerDescriptor<C, S> {
  return function (target, key, descriptor) {
    globalHermesMetadata.addHandler({
      conditions: options.conditions || null,
      decorator: "ViewEventHandler",
      encryption: false,
      handler: descriptor.value,
      key: key.toString(),
      schema: null,
      target: target.constructor as Constructor,
      trigger: EventClass,
    });
  };
}
