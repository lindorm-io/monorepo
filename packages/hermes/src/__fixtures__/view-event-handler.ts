import { ViewStoreType } from "../enums";
import { HermesViewEventHandler } from "../handlers";
import { ViewEventHandlerOptions } from "../types";
import { TEST_AGGREGATE_IDENTIFIER } from "./aggregate";
import {
  TEST_HERMES_EVENT,
  TEST_HERMES_EVENT_CREATE,
  TEST_HERMES_EVENT_DESTROY,
  TEST_HERMES_EVENT_MERGE_STATE,
  TEST_HERMES_EVENT_SET_STATE,
  TEST_HERMES_EVENT_THROWS,
} from "./hermes-event";
import { TEST_VIEW_IDENTIFIER } from "./view";

export const TEST_VIEW_EVENT_HANDLER_OPTIONS: ViewEventHandlerOptions = {
  adapter: { type: ViewStoreType.Custom },
  aggregate: {
    name: TEST_AGGREGATE_IDENTIFIER.name,
    context: TEST_AGGREGATE_IDENTIFIER.context,
  },
  view: {
    name: TEST_VIEW_IDENTIFIER.name,
    context: TEST_VIEW_IDENTIFIER.context,
  },
  eventName: TEST_HERMES_EVENT.name,

  conditions: { created: true },
  getViewId: (event) => event.aggregate.id,
  handler: jest.fn().mockImplementation(async () => {}),
};

export const TEST_VIEW_EVENT_HANDLER = new HermesViewEventHandler(
  TEST_VIEW_EVENT_HANDLER_OPTIONS,
);

export const TEST_VIEW_EVENT_HANDLER_CREATE = new HermesViewEventHandler({
  ...TEST_VIEW_EVENT_HANDLER_OPTIONS,
  eventName: TEST_HERMES_EVENT_CREATE.name,
  conditions: { created: false },
  handler: jest.fn().mockImplementation(async (ctx) => {
    ctx.setState({ created: true });
  }),
});

export const TEST_VIEW_EVENT_HANDLER_DESTROY = new HermesViewEventHandler({
  ...TEST_VIEW_EVENT_HANDLER_OPTIONS,
  eventName: TEST_HERMES_EVENT_DESTROY.name,
  handler: jest.fn().mockImplementation(async (ctx) => {
    ctx.destroy();
  }),
});

export const TEST_VIEW_EVENT_HANDLER_MERGE_STATE = new HermesViewEventHandler({
  ...TEST_VIEW_EVENT_HANDLER_OPTIONS,
  eventName: TEST_HERMES_EVENT_MERGE_STATE.name,
  handler: jest.fn().mockImplementation(async (ctx) => {
    ctx.mergeState({ merge: ctx.event });
  }),
});

export const TEST_VIEW_EVENT_HANDLER_SET_STATE = new HermesViewEventHandler({
  ...TEST_VIEW_EVENT_HANDLER_OPTIONS,
  eventName: TEST_HERMES_EVENT_SET_STATE.name,
  handler: jest.fn().mockImplementation(async (ctx) => {
    ctx.setState({ ...ctx.state, set: ctx.event });
  }),
});

export const TEST_VIEW_EVENT_HANDLER_THROWS = new HermesViewEventHandler({
  ...TEST_VIEW_EVENT_HANDLER_OPTIONS,
  eventName: TEST_HERMES_EVENT_THROWS.name,
  handler: jest.fn().mockImplementation(async () => {
    throw new Error("throw");
  }),
});
