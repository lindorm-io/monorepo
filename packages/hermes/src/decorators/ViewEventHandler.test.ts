import { Dict } from "@lindorm/types";
import { ViewEventCtx } from "../types/handlers/view-event-handler";
import { globalHermesMetadata } from "../utils/private";
import { Aggregate } from "./Aggregate";
import { Event } from "./Event";
import { View } from "./View";
import { ViewEventHandler } from "./ViewEventHandler";

describe("ViewEventHandler Decorator", () => {
  @Aggregate()
  class TestViewAggregate {}

  test("should add metadata", () => {
    @Event()
    class TestViewEventHandlerEvent {}

    @View(TestViewAggregate, "mongo")
    class TestViewEventHandlerView {
      @ViewEventHandler(TestViewEventHandlerEvent)
      public async onTestViewEventHandlerEvent(
        ctx: ViewEventCtx<TestViewEventHandlerEvent, Dict>,
      ) {}
    }

    expect(globalHermesMetadata.getView(TestViewEventHandlerView)).toMatchSnapshot();
  });
});
