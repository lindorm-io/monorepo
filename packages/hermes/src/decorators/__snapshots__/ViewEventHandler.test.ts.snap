// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewEventHandler Decorator should add metadata 1`] = `
{
  "aggregates": [
    [Function],
  ],
  "handlers": [
    {
      "conditions": null,
      "decorator": "ViewEventHandler",
      "encryption": false,
      "handler": [Function],
      "key": "onTestViewEventHandlerEvent",
      "schema": null,
      "target": [Function],
      "trigger": [Function],
    },
  ],
  "name": "test_view_event_handler_view",
  "namespace": null,
  "source": "mongo",
  "target": [Function],
}
`;
